src/probnum/kernels/_kernel.py

```diff
@@ -0,0 +1,273 @@
+"""Kernel / covariance function."""
+
+import abc
+from typing import Generic, Optional, Tuple, TypeVar, Union
+
+import numpy as np
+
+import probnum.utils as _utils
+from probnum.type import IntArgType, ShapeArgType, ShapeType
+
+_InputType = TypeVar("InputType")
+
+
+class Kernel(Generic[_InputType], abc.ABC):
```

Comment:
I think this was a good shout, at least for the foreseeable future. It seems to make the implementation a bit more condensed.